src/main/java/com/limechain/trie/structure/TrieStructure.java

```diff
@@ -1,30 +1,480 @@
+     */
+    @NotNull
+    TrieNode<T> getNodeAtIndexInner(int nodeIndex) {
+        return this.nodes.get(nodeIndex);
+    }
+
+    @Nullable
+    public T getUserDataAtIndex(@NotNull TrieNodeIndex nodeIndex) {
+        return this.getNodeAtIndexInner(nodeIndex.getValue()).userData;
+    }
+
+    @Nullable
+    public NodeHandle<T> nodeHandleAtIndex(TrieNodeIndex index) {
+        return nodeHandleAtIndexInner(index.getValue());
+    }
+
+    /**
+     * This method is a shortcut for {@link TrieStructure#nodeHandleAtIndex(TrieNodeIndex)}
+     * followed by {@link NodeHandle#getFullKey()}
+     *
+     * @param nodeIndex index of a node
+     * @return the key of the node at the given index
+     */
+    @NotNull
+    public Nibbles nodeFullKeyAtIndex(TrieNodeIndex nodeIndex) {
+        return nodeFullKeyAtIndexInner(nodeIndex.getValue());
+    }
+
+    @NotNull
+    Nibbles nodeFullKeyAtIndexInner(int targetNodeIndex) {
+        List<Integer> nodePath = this.nodePath(targetNodeIndex); // path without target itself
+        nodePath.add(targetNodeIndex); // add target to the end
+
+        Stream<Nibble> nibblesStream = nodePath
+            .stream()
+            .flatMap(n -> {
+                TrieNode<T> node = this.getNodeAtIndexInner(n);
+
+                Stream<Nibble> childIndex = Stream.ofNullable(
+                    node.parent == null
+                        ? null
+                        : node.parent.childIndexWithinParent()
+                );
+
+                Nibbles partialKey = node.partialKey.copy();
+
+                return Stream.concat(childIndex, partialKey.stream());
+            });
+
+        return nibblesStream.collect(NibblesCollector.toNibbles());
+    }
+
+    /**
+     * @param targetNodeIndex index of the target node
+     * @return the indices to traverse to reach `target` from root, not including `target` itself.
+     *         So if target is root, returns an empty list.
+     * @throws NullPointerException if targetNodeIndex is not a valid index
+     */
+    List<Integer> nodePath(int targetNodeIndex) {
+        List<Integer> path = new LinkedList<>();
+        var current = this.getNodeAtIndexInner(targetNodeIndex).parent;
+
+        while (current != null) {
+            int nodeIndex = current.parentNodeIndex();
+            path.add(nodeIndex);
+            current = this.getNodeAtIndexInner(nodeIndex).parent;
+        }
+
+        Collections.reverse(path);
+
+        return path;
+
+//        // Alternatively (and more functionally :D):
+//        return Stream.iterate(this.getNodeAtIndexInner(targetNodeIndex).parentIndex,
+//                Objects::nonNull,
+//                index -> this.getNodeAtIndexInner(index.getValue0()).parentIndex)
+//            .map(Pair::getValue0)
+//            .collect(Collectors.collectingAndThen(Collectors.toList(), list -> {
+//                Collections.reverse(list);
+//                return list;
+//            }));
```

Comment:
Is this forgotten or just given as an example of how else it could look like?

Reply:
A little bit of both, but let's say prompting for another opinion was the reason it was not immediately deleted in the initial clean-up. I accept votes on which one is to remain. 😄

Reply:
I'm fine with either, just make sure to delete one of them.